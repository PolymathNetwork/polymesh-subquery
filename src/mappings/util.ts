import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
