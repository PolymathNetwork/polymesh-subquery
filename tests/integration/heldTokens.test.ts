import { gql } from "@apollo/client/core";
import { getApolloClient } from "../util";
const { query } = getApolloClient();

describe("heldTokens", () => {
  test("test getting tokens held by DID", async () => {
    const q = {
      query: gql`
        query {
          heldTokens(
            filter: {
              did: {
                equalTo: "0x0500000000000000000000000000000000000000000000000000000000000000"
              }
            }
          ) {
            totalCount
            nodes {
              token
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});
