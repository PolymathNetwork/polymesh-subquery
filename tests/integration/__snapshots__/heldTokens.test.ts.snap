// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heldTokens test getting tokens held by DID 1`] = `
Object {
  "heldTokens": Object {
    "__typename": "HeldTokensConnection",
    "nodes": Array [
      Object {
        "__typename": "HeldToken",
        "token": "11ATICKER",
      },
      Object {
        "__typename": "HeldToken",
        "token": "11BTICKER1",
      },
      Object {
        "__typename": "HeldToken",
        "token": "11BTICKER2",
      },
      Object {
        "__typename": "HeldToken",
        "token": "12TICKER",
      },
      Object {
        "__typename": "HeldToken",
        "token": "13EARNED",
      },
      Object {
        "__typename": "HeldToken",
        "token": "13TICKER",
      },
      Object {
        "__typename": "HeldToken",
        "token": "14TICKER",
      },
      Object {
        "__typename": "HeldToken",
        "token": "14TICKERRAIS",
      },
    ],
    "totalCount": 8,
  },
}
`;
